CPU speedup

AaronParsons opened the pull request:
Before working on GPU acceleration for linsolve, i thought I'd take another shot at speeding things up on the CPU.

I pulled in various changes using some new numpy functionality (operating on stacks of matrices), eliminating for loops in favor of list comprehensions, and for sparse matrices, rolling out a dictionary-based sparse array format that is effective for when many different A matrices all have the same sparsity structure.

To judge progress, I put in a bunch of new benchmarks to track the impact of speedups.

Finally, there were some nitpicky things that needed to be fixed and I was tweaking code, ranging from specifying tolerances on unit tests to catching when a non-degenerate solver is improperly used on a degenerate system of equations. There were also some slight improvements to handling data types, which were not always handled properly internally.

AaronParsons commented:
Also, here are the results of the speed tests:

BEFORE SPEEDUP:

$ python linsolve/tests/benchmark_A_large_shared.py (DENSE)
Solved in 4.325188398361206

$ python linsolve/tests/benchmark_A_small_shared.py 
Solved in 1.9547922611236572

$ python linsolve/tests/benchmark_A_small_unique.py 
Solved in 6.330256938934326

$ python linsolve/tests/benchmark_A_large_shared.py (SPARSE)
Solved in 1.2443928718566895

$ python linsolve/tests/benchmark_A_small_shared_sparse.py 
KeyboardInterrupt after 1 min

$ python linsolve/tests/benchmark_A_small_unique_sparse.py 
Solved in 40.52456021308899

python linsolve/tests/linsolve_test.py 
.....F..........................................................................
======================================================================
FAIL: test_degen_sol (__main__.TestLinProductSolver)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "linsolve/tests/linsolve_test.py", line 582, in test_degen_sol
    self.assertAlmostEqual(x*y.conjugate(), d['x*y_'], 3)
AssertionError: (4.0265855604517125+0.07777322221213212j) != (4+0j) within 3 places (0.08219164262739453 difference)

----------------------------------------------------------------------
Ran 80 tests in 2.627s

FAILED (failures=1)


AFTER SPEEDUP:

$ python linsolve/tests/benchmark_A_large_shared.py (DENSE)
Solved in 1.6912477016448975
** 2.5x speedup

$ python linsolve/tests/benchmark_A_small_shared.py 
Solved in 0.47655701637268066
** 4x speedup

$ python linsolve/tests/benchmark_A_small_unique.py 
Solved in 4.751166820526123
** 1.3x speedup

$ python linsolve/tests/benchmark_A_large_shared.py (SPARSE)
Solved in 1.203486680984497
** same

$ python linsolve/tests/benchmark_A_small_shared_sparse.py 
Solved in 0.6428780555725098
** > 60x speedup

$ python linsolve/tests/benchmark_A_small_unique_sparse.py 
Solved in 0.4239342212677002
** 95x speedup

$ python linsolve/tests/linsolve_test.py 
................................................................................
----------------------------------------------------------------------
Ran 80 tests in 0.399s

OK
** 6.5x speedup

AaronParsons commented:
@jsdillon I think we need to pull the python2 CI tests, because we should start stripping out support for python2.  Do you know how to do that?

jsdillon commented:
Let me give it a shot...

jsdillon commented:
@AaronParsons I tried, but in the process I discovered that linsolve is broken in python 3.8 due to some changes in `ast` that I don't really understand. Do you want to take a look? See https://travis-ci.org/github/HERA-Team/linsolve/builds/697848505

jsdillon commented:
Ok, figured it out. #29 is ready for your review @AaronParsons

AaronParsons commented:
@jsdillon Ok, merged in #29 and now all CI tests pass. The rest is ready for your review, I think.

jsdillon commented:
I have verified that hera_cal tests still pass with this PR. The time it takes to run `redcal`'s tests has not changed appreciably.

jsdillon commented on linsolve/linsolve.py:
```diff
@@ -250,6 +251,7 @@ def __init__(self, data, wgts={}, sparse=False, **kwargs):
         Returns:
```
I think the description of the kwarg `sparse` is no longer correct and we should update it to clarify that while the A matrices are represented sparsely, the AtA matrices are not.

jsdillon commented on linsolve/linsolve.py:
```diff
@@ -568,14 +612,18 @@ def __init__(self, data, wgts={}, sparse=False, **kwargs):
             logamp_consts[k], logphs_consts[k] = c.real, c.imag
         self.ls_amp = LinearSolver(logamp, logampw, sparse=sparse, constants=logamp_consts)
-        self.ls_phs = LinearSolver(logphs, logphsw, sparse=sparse, constants=logphs_consts)
+        if self.dtype in (np.complex64, np.complex128): # XXX worry about enumrating these explicitly
```
This is a bit hacky, but what if we do `if np.iscomplexobj(np.zeros(1, dtype=self.dtype))`? There might be another way involve subtypes, but I'm not sure.

AaronParsons replied:
I think this is best done as an overhaul of how data types are handled (which should be made explicit) in #33. Let's leave this as tbd for now. As implemented, I have not seen it break anything.

jsdillon commented on linsolve/tests/benchmark_A_large_shared.py:
```diff
@@ -0,0 +1,28 @@
+import linsolve
```
I'd be good to say in a comment at the top what these various benchmarks are testing (just to be more explicit than the filenames of the python scripts).

AaronParsons commented on linsolve/linsolve.py:
```diff
@@ -250,6 +251,7 @@ def __init__(self, data, wgts={}, sparse=False, **kwargs):
         Returns:
             None
         """
+        # XXX add ability to override datatype inference
```
I like having them in code, b/c they end up localized to certain areas where, if I'm editing, it will get my attention. However, I will see you halfway and make sure each is an issue, and that the issue is linked in the code.

jsdillon replied:
Fine with me!

AaronParsons commented on linsolve/linsolve.py:
```diff
@@ -348,87 +355,130 @@ def get_weighted_data(self):
                 or (if sparse) as atol and btol in scipy.sparse.linalg.lsqr
+                Default: None (resolves to machine precision for inferred dtype)
             mode: 'default', 'lsqr', 'pinv', or 'solve', selects which inverter to use. 
                 'default': tries 'lsqr' but if a LinAlgError is encountered, backs off to try 'pinv'.
```
Don't understand the "default" comment above. I had changed docstring here based on the changes I made at line 504.

jsdillon replied:
Sorry, I must have misunderstood. That said, as it stands, `default` and `pinv` mean the same thing. I'd probably add that to the docstring.

AaronParsons replied:
Ah, I see what you mean now. I think github was tagging the wrong line. Fixed.

AaronParsons commented on linsolve/linsolve.py:
```diff
@@ -446,37 +496,31 @@ def solve(self, rcond=1e-10, mode='default', verbose=False):
+                x = _invert(A, y, rcond)
 
         x.shape = x.shape[:1] + self._data_shape # restore to shape of original data
         sol = {}
```
Sorry, it not immediately obvious to me what you are referring to here. Are you suggesting the if/else stuff could be a dictionary comprehension, or ...?

jsdillon replied:
I guess it's out of the scope of the PR, but I noticed that the construction of the `sol` dict could be a dictionary comprehension. Don't feel strongly.

AaronParsons replied:
I see what you mean now (github was tagging the wrong lines). I'm not inclined to change something that's working and outside the scope. I presume there was a reason it was done this way, given the extensive use of dictcomps elsewhere.

AaronParsons commented:
@jsdillon Just pushed a commit addressing most of your feedback. There was only a couple of things that I couldn't address, which I responded to above. LMK what you think.
